File: stock_fetcher.js
/**
 * Stock Data Fetcher - Fetches stock data and calculates basic technical indicators
 * Usage: node stock_fetcher.js SYMBOL DAYS
 * Example: node stock_fetcher.js AAPL 30
 */

const https = require('https');
const fs = require('fs');

class StockFetcher {
    constructor() {
        this.baseUrl = 'https://query1.finance.yahoo.com/v8/finance/chart/';
    }

    async fetchStockData(symbol, days = 30) {
        return new Promise((resolve, reject) => {
            const period1 = Math.floor(Date.now() / 1000) - (days * 24 * 60 * 60);
            const period2 = Math.floor(Date.now() / 1000);
            
            const url = `${this.baseUrl}${symbol}?period1=${period1}&period2=${period2}&interval=1d`;
            
            https.get(url, (response) => {
                let data = '';
                
                response.on('data', (chunk) => {
                    data += chunk;
                });
                
                response.on('end', () => {
                    try {
                        const result = JSON.parse(data);
                        const stockData = this.processData(result, symbol);
                        resolve(stockData);
                    } catch (error) {
                        reject(new Error('Failed to parse stock data'));
                    }
                });
                
            }).on('error', (error) => {
                reject(error);
            });
        });
    }

    processData(data, symbol) {
        if (!data.chart || !data.chart.result || !data.chart.result[0]) {
            throw new Error('Invalid stock data received');
        }

        const result = data.chart.result[0];
        const timestamps = result.timestamp;
        const quotes = result.indicators.quote[0];
        
        const stockData = [];
        
        for (let i = 0; i < timestamps.length; i++) {
            if (quotes.open[i] && quotes.high[i] && quotes.low[i] && quotes.close[i]) {
                stockData.push({
                    date: new Date(timestamps[i] * 1000).toISOString().split('T')[0],
                    open: quotes.open[i],
                    high: quotes.high[i],
                    low: quotes.low[i],
                    close: quotes.close[i],
                    volume: quotes.volume[i] || 0
                });
            }
        }
        
        // Calculate technical indicators
        const dataWithIndicators = this.calculateIndicators(stockData);
        
        return {
            symbol: symbol,
            data: dataWithIndicators
        };
    }

    calculateIndicators(data) {
        // Calculate Simple Moving Average (SMA)
        for (let i = 0; i < data.length; i++) {
            // SMA 20
            if (i >= 19) {
                let sum = 0;
                for (let j = i - 19; j <= i; j++) {
                    sum += data[j].close;
                }
                data[i].sma_20 = sum / 20;
            }
            
            // SMA 50
            if (i >= 49) {
                let sum = 0;
                for (let j = i - 49; j <= i; j++) {
                    sum += data[j].close;
                }
                data[i].sma_50 = sum / 50;
            }
            
            // RSI (simplified calculation)
            if (i > 0) {
                const change = data[i].close - data[i-1].close;
                data[i].price_change = change;
            }
        }
        
        // Calculate RSI (14-period)
        for (let i = 14; i < data.length; i++) {
            let gains = 0;
            let losses = 0;
            
            for (let j = i - 13; j <= i; j++) {
                if (data[j].price_change > 0) {
                    gains += data[j].price_change;
                } else {
                    losses += Math.abs(data[j].price_change);
                }
            }
            
            const avgGain = gains / 14;
            const avgLoss = losses / 14;
            const rs = avgGain / avgLoss;
            data[i].rsi = 100 - (100 / (1 + rs));
        }
        
        return data;
    }

    saveToJSON(data, filename = 'stock_data.json') {
        fs.writeFileSync(filename, JSON.stringify(data, null, 2));
        console.log(`Data saved to ${filename}`);
    }
}

// Main execution
if (require.main === module) {
    const args = process.argv.slice(2);
    
    if (args.length < 1) {
        console.log('Usage: node stock_fetcher.js <SYMBOL> [DAYS]');
        console.log('Example: node stock_fetcher.js AAPL 30');
        process.exit(1);
    }
    
    const symbol = args[0].toUpperCase();
    const days = args[1] ? parseInt(args[1]) : 30;
    
    const fetcher = new StockFetcher();
    
    console.log(`Fetching ${days} days of data for ${symbol}...`);
    
    fetcher.fetchStockData(symbol, days)
        .then(data => {
            console.log(`Successfully fetched ${data.data.length} days of data for ${symbol}`);
            fetcher.saveToJSON(data);
        })
        .catch(error => {
            console.error('Error:', error.message);
        });
}

module.exports = StockFetcher;

File: stock_fetcher.py
#!/usr/bin/env python3
"""
Stock Data Fetcher - Fetches historical stock data from Alpha Vantage API
"""

import requests
import pandas as pd
import json
import sys
from datetime import datetime, timedelta

class StockFetcher:
    def __init__(self, api_key=None):
        # Use demo key or provide your own Alpha Vantage API key
        self.api_key = api_key or "demo"
        self.base_url = "https://www.alphavantage.co/query"
    
    def fetch_stock_data(self, symbol, period="3month"):
        """
        Fetch historical stock data for a given symbol
        Periods: 1month, 3month, 1year, 2year
        """
        try:
            # Map periods to Alpha Vantage function
            period_map = {
                "1month": "TIME_SERIES_DAILY",
                "3month": "TIME_SERIES_DAILY", 
                "1year": "TIME_SERIES_WEEKLY",
                "2year": "TIME_SERIES_WEEKLY"
            }
            
            function = period_map.get(period, "TIME_SERIES_DAILY")
            
            params = {
                "function": function,
                "symbol": symbol,
                "apikey": self.api_key,
                "outputsize": "compact" if period == "1month" else "full"
            }
            
            response = requests.get(self.base_url, params=params)
            data = response.json()
            
            if "Error Message" in data:
                return {"error": f"Invalid symbol: {symbol}"}
            
            # Extract time series data
            time_series_key = "Time Series (Daily)" if "Daily" in function else "Weekly Time Series"
            
            if time_series_key not in data:
                return {"error": f"No data found for symbol: {symbol}"}
            
            time_series = data[time_series_key]
            
            # Convert to DataFrame and calculate technical indicators
            df = self._process_data(time_series, symbol)
            return df.to_dict('records')
            
        except Exception as e:
            return {"error": f"Error fetching data: {str(e)}"}
    
    def _process_data(self, time_series, symbol):
        """Process raw time series data and calculate technical indicators"""
        records = []
        
        for date, values in time_series.items():
            records.append({
                "date": date,
                "symbol": symbol,
                "open": float(values["1. open"]),
                "high": float(values["2. high"]),
                "low": float(values["3. low"]),
                "close": float(values["4. close"]),
                "volume": int(values["5. volume"])
            })
        
        df = pd.DataFrame(records)
        df['date'] = pd.to_datetime(df['date'])
        df = df.sort_values('date')
        
        # Calculate technical indicators
        df = self._calculate_technical_indicators(df)
        return df
    
    def _calculate_technical_indicators(self, df):
        """Calculate basic technical indicators"""
        # Simple Moving Averages
        df['sma_20'] = df['close'].rolling(window=20).mean()
        df['sma_50'] = df['close'].rolling(window=50).mean()
        
        # RSI (Relative Strength Index)
        df['rsi'] = self._calculate_rsi(df['close'])
        
        # MACD
        df['macd'], df['macd_signal'] = self._calculate_macd(df['close'])
        
        # Bollinger Bands
        df['bb_upper'], df['bb_lower'] = self._calculate_bollinger_bands(df['close'])
        
        return df
    
    def _calculate_rsi(self, prices, period=14):
        """Calculate RSI indicator"""
        delta = prices.diff()
        gain = (delta.where(delta > 0, 0)).rolling(window=period).mean()
        loss = (-delta.where(delta < 0, 0)).rolling(window=period).mean()
        rs = gain / loss
        rsi = 100 - (100 / (1 + rs))
        return rsi
    
    def _calculate_macd(self, prices, fast=12, slow=26, signal=9):
        """Calculate MACD indicator"""
        ema_fast = prices.ewm(span=fast).mean()
        ema_slow = prices.ewm(span=slow).mean()
        macd = ema_fast - ema_slow
        macd_signal = macd.ewm(span=signal).mean()
        return macd, macd_signal
    
    def _calculate_bollinger_bands(self, prices, period=20, std_dev=2):
        """Calculate Bollinger Bands"""
        sma = prices.rolling(window=period).mean()
        std = prices.rolling(window=period).std()
        upper_band = sma + (std * std_dev)
        lower_band = sma - (std * std_dev)
        return upper_band, lower_band

def main():
    if len(sys.argv) < 2:
        print("Usage: python stock_fetcher.py <symbol> [period]")
        print("Example: python stock_fetcher.py AAPL 3month")
        sys.exit(1)
    
    symbol = sys.argv[1]
    period = sys.argv[2] if len(sys.argv) > 2 else "3month"
    
    fetcher = StockFetcher()
    data = fetcher.fetch_stock_data(symbol, period)
    
    if "error" in data:
        print(f"Error: {data['error']}")
        sys.exit(1)
    
    # Save data to JSON file for JavaScript to use
    output_file = f"{symbol.lower()}_data.json"
    with open(output_file, 'w') as f:
        json.dump(data, f, indent=2, default=str)
    
    print(f"Data saved to {output_file}")
    print(f"Fetched {len(data)} records for {symbol}")

if __name__ == "__main__":
    main()